Section 3.1 - цикл while

aleno4ka1998 opened the pull request:
Мне кажется лучше поделить задачи 3 секции на задачи по for и while, а то их довольно много и там и там

aleno4ka1998 committed:
Список квадратов

Выведите все точные квадраты натуральных чисел, не превосходящие данного числа N.

Входные данные
Задано единственное целое число N

Выходные данные
Необходимо вывести  все точные квадраты натуральных чисел, не превосходящие данного числа N.

aleno4ka1998 committed:
Минимальный делитель

Дано целое число, не меньшее 2. Выведите его наименьший натуральный делитель, отличный от 1.

Входные данные
Вводится натуральное число.

Выходные данные
Выведите ответ на задачу.

aleno4ka1998 committed:
Список степеней двойки

По данному числу N распечатайте все целые степени двойки, не превосходящие N, в порядке возрастания.

Операцией возведения в степень пользоваться нельзя!

Входные данные
Вводится натуральное число.

Выходные данные
Выведите ответ на задачу.

aleno4ka1998 committed:
Точная степень двойки

Дано натуральное число N. Выведите слово YES, если число N является точной степенью двойки, или слово NO в противном случае.

Операцией возведения в степень пользоваться нельзя!

aleno4ka1998 committed:
Двоичный логарифм

По данному натуральному числу N выведите такое наименьшее целое число k, что 2k≥N.

Операцией возведения в степень пользоваться нельзя!

aleno4ka1998 committed:
Утренняя пробежка

В первый день спортсмен пробежал x километров, а затем он каждый день увеличивал пробег на 10% от предыдущего значения. По данному числу y определите номер дня, на который пробег спортсмена составит не менее y километров.

Входные данные
Программа получает на вход действительные числа x и y

Выходные данные
Программа должна вывести одно натуральное число.

aleno4ka1998 committed:
Банковские проценты

Вклад в банке составляет x рублей. Ежегодно он увеличивается на p процентов, после чего дробная часть копеек отбрасывается. Каждый год сумма вклада становится больше. Определите, через сколько лет вклад составит не менее y рублей.

Входные данные
Программа получает на вход три натуральных числа: x, p, y.

Выходные данные
Программа должна вывести одно целое число.

aleno4ka1998 committed:
Числа Фибоначчи

Последовательность Фибоначчи определяется так:

φ0=0,  φ1=1, ..., φn=φn-1+φn-2.

По данному числу n определите n-е число Фибоначчи φn.

aleno4ka1998 committed:
Номер числа Фибоначчи

Дано натуральное число A > 1. Определите, каким по счету числом Фибоначчи оно является, то есть выведите такое число n, что φn=A. Если А не является числом Фибоначчи, выведите число -1.

aleno4ka1998 committed:
Исполнитель "Раздвоитель"

Исполнитель “Раздвоитель” преобразует натуральные числа. У него есть две команды: “Вычесть 1” и “Разделить на 2”, первая команда уменьшает число на 1, вторая команда уменьшает число в два раза, если оно чётное, иначе происходит ошибка.

Входные данные
Дано два натуральных числа A и B (A>B).

Выходные данные
Напишите алгоритм для Развоителя, который преобразует число A в число B и при этом содержит минимальное число команд. Команды алгоритма нужно выводить по одной в строке, первая команда обозначается, как -1, вторая команда как :2.

aleno4ka1998 committed:
Длина последовательности

Программа получает на вход последовательность целых неотрицательных чисел, каждое число записано в отдельной строке. Последовательность завершается числом 0, при считывании которого программа должна закончить свою работу и вывести количество членов последовательности (не считая завершающего числа 0).

Числа, следующие за числом 0, считывать не нужно.

aleno4ka1998 committed:
Сумма последовательности

Определите сумму всех элементов последовательности, завершающейся числом 0.

Числа, следующие за нулем, считывать не нужно.

Входные данные
Вводится последовательность целых чисел, оканчивающаяся числом 0 (само число 0 в последовательность не входит).

Выходные данные
Выведите ответ на задачу.

aleno4ka1998 committed:
Среднее значение последовательности

Определите среднее арифметическое элементов последовательности, завершающейся числом 0.

Число 0 в последовательность не входит. Числа, следующие за нулем, считывать не нужно.

Входные данные
Вводится последовательность целых чисел. Ввод завершается, когда будет введено число 0.

Выходные данные
Выведите одно число - среднее арифметическое элементов последовательности

aleno4ka1998 committed:
Количество чётных элементов последовательности

Определите количество четных элементов в последовательности, завершающейся числом 0.

Само число 0, и все, что следует за ним, учитывать не нужно.

Входные данные
Вводится последовательность целых чисел, оканчивающаяся числом 0 (само число 0 в последовательность не входит).

Выходные данные
Выведите ответ на задачу.

aleno4ka1998 committed:
Максимум последовательности

Последовательность состоит из натуральных чисел и завершается числом 0. Определите значение наибольшего элемента последовательности.

Числа, следующие за нулем, считывать не нужно.

Входные данные
Вводится последовательность целых чисел, оканчивающаяся числом 0 (само число 0 в последовательность не входит).

Выходные данные
Выведите ответ на задачу.

aleno4ka1998 committed:
Количество элементов, которые больше предыдущего

Последовательность состоит из натуральных чисел и завершается числом 0. Определите, сколько элементов этой последовательности больше предыдущего элемента.

Числа, следующие за числом 0, считывать не нужно.

Входные данные
Вводится последовательность целых чисел, оканчивающаяся числом 0 (само число 0 в последовательность не входит).

Выходные данные
Выведите ответ на задачу.

thelightningseas commented on section3_1/average sequence_number.py:
```diff
@@ -0,0 +1,9 @@
+while a != 0:
+    a = int(input())
+    k += a
+    if a != 0:
```
а можно было бы тут обойтись без двух проверок, что a не равен 0?

aleno4ka1998 replied:
да, можно объявить первый инпут перед циклом, а в самом цикле поменять местами расчёты и инпут

thelightningseas commented on section3_1/bank_interest.py:
```diff
@@ -0,0 +1,8 @@
+p = int(input())
+y = int(input())
+k = 0
+while x < y:
```
а понятно как сделать без цикла такую задачу? в одну строчку

aleno4ka1998 replied:
Можно
x = int(input())
p = int(input())
y = int(input())
k = round(y / (int(x + x / 100 * p)))
print(k)

thelightningseas commented on section3_1/fibonacci.py:
```diff
@@ -0,0 +1,10 @@
+b = 1
+c = 0
+k = 1
+while k != n:
```
кажется, конструкция ```while k != n: k += 1``` эквивалентна ```for _ in range(n)```, но выглядит более громоздко

aleno4ka1998 replied:
все время кажется что если задача в рамках раздела "Задачи на цикл while", то нужно использовать цикл while)

thelightningseas commented on section3_1/max_of_sequence.py:
```diff
@@ -0,0 +1,9 @@
+k = 0
+maximum = 0
+while a != 0:
+    k = a
```
а зачем тут ```k```?

aleno4ka1998 replied:
Прошу прощения, прога нерабочая. Вот, исправляюсь:
```
a = int(input())
maximum = 0
while a != 0:
    if a > maximum:
        maximum = a
    a = int(input())
print(maximum)
```

thelightningseas commented on section3_1/sequence_length.py:
```diff
@@ -0,0 +1,7 @@
+n = 1
+k = 0
+while n != 0:
```
опять очень перегруженно

```
sequence_length = 0
while True:
    next_num = int(input())
    if next_num == 0:
        break
    sequence_length += 1
print(sequence_length)
```

aleno4ka1998 replied:
while True продолжает цикл до события вызывающего break?

thelightningseas replied:
да. или до бесконечности если его не будет)

вообще ```while condition``` работает как раз так ведь
смотрит на ```condition``` которы там написан (например ```n != 0```) если то что в нем написано - ```True```, делает еще итерацию, если ```False``` - останавливается. вайл тру просто всегда верен

aleno4ka1998 committed:
Второй максимум

Последовательность состоит из различных натуральных чисел и завершается числом 0. Определите значение второго по величине элемента в этой последовательности.

Числа, следующие за числом 0, считывать не нужно.

Входные данные
Вводится последовательность целых чисел, оканчивающаяся числом 0 (само число 0 в последовательность не входит).

Выходные данные
Выведите ответ на задачу.

aleno4ka1998 committed:
Revert "Второй максимум"

This reverts commit 4e75dcd8a440b70842270301286daecbacb03981.

aleno4ka1998 committed:
Второй максимум

Последовательность состоит из различных натуральных чисел и завершается числом 0. Определите значение второго по величине элемента в этой последовательности.

Числа, следующие за числом 0, считывать не нужно.

aleno4ka1998 committed:
Количество элементов, равных максимуму

Последовательность состоит из натуральных чисел и завершается числом 0. Всего вводится не более 10000 чисел (не считая завершающего числа 0). Определите, сколько элементов этой последовательности равны ее наибольшему элементу.

Числа, следующие за числом 0, считывать не нужно.

aleno4ka1998 committed:
Сумма последовательности

Найдите сумму последовательности натуральных чисел, если признаком окончания конца последовательности является два подряд идущих числа 0.

Числа, следующие после двух подряд идущих нулей считывать не нужно.

Входные данные
Вводится последовательность целых чисел, оканчивающаяся числом 0 (само число 0 в последовательность не входит).

Выходные данные
Выведите ответ на задачу.

Примеры
входные данные
1
0
7
0
9
0
0
5
выходные данные
17

aleno4ka1998 committed:
Максимальное число идущих подряд равных элементов

Дана последовательность натуральных чисел, завершающаяся числом 0. Определите, какое наибольшее число подряд идущих элементов этой последовательности равны друг другу.

Числа, следующие за числом 0, считывать не нужно.

thelightningseas commented on section3_1/quantity of maximum numbers.py:
```diff
@@ -0,0 +1,10 @@
+a = int(input())
+maximum = 0
```
ну тут решение не верное)

aleno4ka1998 replied:
Почему? Сначала находим первый элемент-максимум, потом равные ему

thelightningseas commented on section3_1/sequence_sum_2.py:
```diff
@@ -0,0 +1,13 @@
+a = 0
```
очень запутано. лучше что то такое:
```
sequence_sum = 0
last_number_is_zero = False

while True:
    next_number = int(input())
    if next_number == 0 and last_number_is_zero:
        break
    elif next_number == 0:
        last_number_is_zero = True
    else:
        sequence_sum += next_number
        last_number_is_zero = False

print(sequence_sum)
```

aleno4ka1998 committed:
Исправленная "Количество элементов, равных максимуму"

Добавлено обнуление счётчика при обнаружении нового максимума

aleno4ka1998 committed:
Максимальная длина монотонного фрагмента

Дана последовательность натуральных чисел, завершающаяся число 0. Определите наибольшую длину монотонного фрагмента последовательности (то есть такого фрагмента, где все элементы либо больше предыдущего, либо меньше).

Числа, следующие за числом 0, считывать не нужно.

aleno4ka1998 committed:
Количество локальных максимумов

Элемент последовательности называется локальным максимумом, если он строго больше предыдущего и последующего элемента последовательности. Первый и последний элемент последовательности не являются локальными максимумами.

Дана последовательность натуральных чисел, признаком конца которой является число 0. Определите количество строгих локальных максимумов в этой последовательности.

aleno4ka1998 committed:
Сумма цифр чисел

Дано натуральное число N. Напишите функцию int SumOfDigits (int n) (C/C++), function SumOfDigits (n:longint):integer (Pascal), вычисляющую сумму цифр числа N.

Входные данные
Задано единственное число N

Выходные данные
Необходимо вывести  сумму цифр числа N.

aleno4ka1998 committed:
Количество нолей

Дано натуральное число N. Напишите функцию int NumberOfZeroes(int n) (C/C++/Java), function NumberOfZeroes(n: longint): integer (Pascal), определяющую количество нулей среди всех цифр числа N.

Входные данные
Задано единственное число N

Выходные данные
Необходимо вывести количество нулей среди всех цифр числа N.

aleno4ka1998 committed:
Минимальная и максимальная цифры

Дано натуральное число N. Напишите функцию int MinDigit (int n) (C/C++), function MinDigit (n:longint):integer (Pascal) и int MaxDigit (int n) (C/C++), function MaxDigit (n:longint):integer (Pascal), определяющую наименьшую и наибольшую цифры данного числа.

Входные данные
Задано единственное число N

Выходные данные
Необходимо вывести наименьшую и наибольшую цифры данного числа через пробел.

aleno4ka1998 committed:
Двоичная запись

Дано натуральное число N. Выведите его представление в двоичном виде в обратном порядке.

Входные данные
Задано единственное число N

Выходные данные
Необходимо вывести требуемое представление числа N.

aleno4ka1998 committed:
Обращение числа

Напишите функцию int reverse(int n) (C/C++), function reverse (n:integer):integer (Pascal), которая переставляет цифры числа в обратном порядке .

Входные данные
Задано единственное число N

Выходные данные
Необходимо вывести цифры данного числа в обратном порядке.

aleno4ka1998 committed:
Количество палиндромов

Назовем число палиндромом, если оно не меняется при перестановке его цифр в обратном порядке. Напишите программу, которая по заданному числу K выводит количество натуральных палиндромов, не превосходящих K.

Входные данные
Задано единственное число K (1≤K≤100000).

Выходные данные
Необходимо вывести количество натуральных палиндромов, не превосходящих K.

thelightningseas commented on section3_1/minimum and maximum.py:
```diff
@@ -0,0 +1,24 @@
+def max_digit(n: int):
+    max_d = 0
```
```max_d``` не очень хорошо. лучше писать целиком ```max_digit```. Но в нашем случае, когда из названия функции понятно, что она вернет, можно назвать например ```result``` и будет хорошо

thelightningseas replied:
и если пишешь тайп хинтинги, то лучше возвращаемое значение тоже указывать
```
def max_digit(n: int) -> int:
   #  some code here
```

thelightningseas commented on section3_1/number of polyindromes.py:
```diff
@@ -0,0 +1,16 @@
+def reverse(n):
+    rev = str('')
```
слишком сложная инициализация строки)
лучше или просто ```rev = str()``` или ```rev = ''```
